### PR TITLE
Sign up done inshallah

### PR DESCRIPTION
**Pull request by aabdoo23**
#### PR Classification
New feature: Introduced MediatR for handling student creation and set up the database schema.

#### PR Summary
This PR introduces MediatR for handling student creation, sets up the database schema, and updates project configurations.
- Added `StudentsController` in `StudentController.cs` using MediatR for student creation.
- Created initial migration files to set up the database schema.
- Updated `Program.cs` to register new services for persistence, infrastructure, and MediatR.
- Added repository and service interfaces and implementations for `Admin`, `Major`, `Moderator`, and `Student` entities.
- Configured `ForumsDbContext` and added `DbContextRegistrationService` for PostgreSQL database integration.

### CODE REVIEW

File: SchedsForums.Api/Controllers/StudentController.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+{
+    [ApiController]
+    [Route("api/[controller]")]
+    public class StudentsController : ControllerBase
+    {
+        private readonly IMediator _mediator;
+
+        public StudentsController(IMediator mediator)
+        {
+            _mediator = mediator;
```
Comment (MostafaHgz) [inline SchedsForums.Api/Controllers/StudentController.cs]: add null check 
`_mediator = mediator ?? throw new ArgumentNullAgreement(nameof(mediator));`

File: SchedsForums.Api/Program.cs [hunk 2]
```diff
@@ -7,13 +11,19 @@ public static void Main(string[] args)
         {
             var builder = WebApplication.CreateBuilder(args);
 
-            // Add services to the container.
```
Comment (MostafaHgz) [inline SchedsForums.Api/Program.cs]: remove extra line

File: SchedsForums.Application/Commands/Students/Create/CreateStudentCommandHandler.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+using SchedsForums.Domain.Interfaces.Repositories;
+using SchedsForums.Infrastructure.Services.Interfaces;
+
+namespace SchedsForums.Application.Commands.Students.Create
+{
+    public class CreateStudentCommandHandler : IRequestHandler<CreateStudentCommand, StudentDTO>
+    {
+        private readonly IStudentRepository _studentRepository;
+        private readonly IPasswordService _passwordService;
+        private readonly IValidator<CreateStudentDTO> _validator = new CreateStudentValidator();
```
Comment (MostafaHgz) [inline SchedsForums.Application/Commands/Students/Create/CreateStudentCommandHandler.cs]: never use new in .net. we mostly use dependency injection for constructing any new object.

File: SchedsForums.Application/Commands/Students/Create/CreateStudentCommandHandler.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+        public CreateStudentCommandHandler(IStudentRepository studentRepository, IPasswordService passwordService)
+        {
+            _studentRepository = studentRepository;
+            _passwordService = passwordService;
+        }
+
+        public async Task<StudentDTO> Handle(CreateStudentCommand request, CancellationToken cancellationToken)
+        {
+            var validationResult = await _validator.ValidateAsync(request.Student);
```
Comment (MostafaHgz) [inline SchedsForums.Application/Commands/Students/Create/CreateStudentCommandHandler.cs]: from 24-28 will be always repeated in all commands/queries. and we need to keep the code dry
look up `AddValidatorFromAssembly()` in fluent validation docs

File: SchedsForums.Application/RegisterServices.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+namespace SchedsForums.Application
+{
+    public static class RegisterServices
+    {
+        public static IServiceCollection AddMediatRServices(this IServiceCollection services)
+        {
+            services.AddMediatR(options => options.RegisterServicesFromAssemblies(typeof(CreateStudentCommandHandler).Assembly));
+
+            return services;
+
```
Comment (MostafaHgz) [inline SchedsForums.Application/RegisterServices.cs:15]: remove extra line

File: SchedsForums.Domain/Entities/Common/BaseUser.cs [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+        protected BaseUser(string name, string email, string password) : base()
+        {
+            Name = name;
+            Email = email;
+            Password = password;
+        }
+        protected BaseUser()
+        {
+        }
+
```
Comment (MostafaHgz) [inline SchedsForums.Domain/Entities/Common/BaseUser.cs]: remove extra line

File: SchedsForums.Domain/Interfaces/IGenericRepository.cs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+﻿namespace SchedsForums.Domain.Interfaces
+{
+    public interface IGenericRepository<T> where T : IBaseEntity
+    {
+        Task<T> GetByIdAsync(string id);
+        Task<IEnumerable<T>> GetAllAsync();
+        Task<T> InsertAsync(T entity);
+        Task<T> UpdateAsync(T entity);
+        Task<T> DeleteAsync(string id);
+
```
Comment (MostafaHgz) [inline SchedsForums.Domain/Interfaces/IGenericRepository.cs]: remove extra line

File: SchedsForums.Domain/ServiceConfiguration/ConfigureServices.cs [hunk 8]
```diff
@@ -0,0 +1,13 @@
+﻿using Microsoft.Extensions.DependencyInjection;
+
+namespace SchedsForums.Domain.ServiceConfiguration
+{
+    public static class ConfigureServices
+    {
+        public static IServiceCollection Configure(this IServiceCollection services)
+        {
+
```
Comment (MostafaHgz) [inline SchedsForums.Domain/ServiceConfiguration/ConfigureServices.cs]: remove extra line

File: SchedsForums.Domain/ServiceConfiguration/ConfigureServices.cs [hunk 9]
```diff
@@ -0,0 +1,13 @@
+﻿using Microsoft.Extensions.DependencyInjection;
+
+namespace SchedsForums.Domain.ServiceConfiguration
+{
+    public static class ConfigureServices
+    {
+        public static IServiceCollection Configure(this IServiceCollection services)
```
Comment (MostafaHgz) [inline SchedsForums.Domain/ServiceConfiguration/ConfigureServices.cs]: what does this do?

File: SchedsForums.Domain/ServiceConfiguration/ConfigureValidators.cs [hunk 10]
```diff
@@ -0,0 +1,14 @@
+﻿using Microsoft.Extensions.DependencyInjection;
+
+namespace SchedsForums.Domain.ServiceConfiguration
+{
+    public static class ConfigureValidators
+    {
+        public static IServiceCollection Configure(this IServiceCollection services)
```
Comment (MostafaHgz) [inline SchedsForums.Domain/ServiceConfiguration/ConfigureValidators.cs]: you now have 2 extensions methods to IServiceCollection that are named the same. name the method itself a meaningful name
also this file does nothing

File: SchedsForums.Domain/Validators/BaseUserValidator.cs [hunk 11]
```diff
@@ -0,0 +1,26 @@
+﻿using FluentValidation;
+using SchedsForums.Domain.Entities.Common;
+
+namespace SchedsForums.Domain.Validators
+{
+    public class BaseUserValidator<T> : AbstractValidator<T> where T : BaseUser
+    {
+        public BaseUserValidator()
```
Comment (MostafaHgz) [inline SchedsForums.Domain/Validators/BaseUserValidator.cs]: this validation is repeated in user validation as well. see how to use common validators in fluent validations

File: SchedsForums.Infrastructure/Contexts/ForumsDbContext.cs [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+namespace SchedsForums.Infrastructure.Contexts
+{
+    public class ForumsDbContext : DbContext
+    {
+        public ForumsDbContext(DbContextOptions<ForumsDbContext> options) : base(options)
+        {
+        }
+        protected override void OnModelCreating(ModelBuilder modelBuilder)
+        {
+            modelBuilder.Entity<BaseUser>()
```
Comment (MostafaHgz) [inline SchedsForums.Infrastructure/Contexts/ForumsDbContext.cs]: when you get to have 10 entities in the future this method will become 100 lines and will be a nightmare to read. 
check how to modularize model configuration and have a configuration file for each entity

File: SchedsForums.Infrastructure/Contexts/ForumsDbContext.cs [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+                .HasValue<Moderator>("Moderator");
+
+            base.OnModelCreating(modelBuilder);
+        }
+
+        public DbSet<BaseUser> Users { get; set; }
+        public DbSet<Admin> Admins { get; set; }
+        public DbSet<Student> Students { get; set; }
+        public DbSet<Moderator> Moderators { get; set; }
+
```
Comment (MostafaHgz) [inline SchedsForums.Infrastructure/Contexts/ForumsDbContext.cs]: class properties always come first before any methods or constructors
also 2 extra lines here

File: SchedsForums.Infrastructure/Repositories/AdminRepository.cs [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+using SchedsForums.Infrastructure.Repositories.Common;
+
+namespace SchedsForums.Infrastructure.Repositories
+{
+    public class AdminRepository : GenericRepository<Admin>, IAdminRepository
+    {
+        private readonly ForumsDbContext _context;
+        public AdminRepository(ForumsDbContext context) : base(context)
+        {
+            _context = context;
```
Comment (MostafaHgz) [inline SchedsForums.Infrastructure/Repositories/AdminRepository.cs]: add null check for the context here

File: SchedsForums.Domain/Entities/Common/BaseUser.cs [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+        public string Email { get; set; }
+        public string Password { get; set; }
+
+        protected BaseUser(string name, string email, string password) : base()
+        {
+            Name = name;
+            Email = email;
+            Password = password;
+        }
+        protected BaseUser()
```
Comment (MostafaHgz) [inline SchedsForums.Domain/Entities/Common/BaseUser.cs]: remove all these empty constructors as they do nothing

File: SchedsForums.Infrastructure/Repositories/Common/GenericRepository.cs [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+        public async Task<T> InsertAsync(T entity)
+        {
+            await _dbSet.AddAsync(entity);
+            await _context.SaveChangesAsync();
+            return entity;
+        }
+
+        public async Task<T> DeleteAsync(string id)
+        {
+            var entity = await _dbSet.FindAsync(id) ?? throw new NullReferenceException($"Can't find an entity with this Id: {id}.");
```
Comment (MostafaHgz) [inline SchedsForums.Infrastructure/Repositories/Common/GenericRepository.cs]: this is not null reference exception. this is when something that cannot be null 
also it's not the responsibility of the repository to throw errors. 
this method requires 2 query calls and it only needs one. see how you can delete with one query

File: SchedsForums.Infrastructure/Repositories/Common/GenericRepository.cs [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+        }
+
+        public async Task<IEnumerable<T>> GetAllAsync()
+        {
+            return await _dbSet.ToListAsync();
+        }
+
+        public async Task<T> GetByIdAsync(string id)
+        {
+            return await _dbSet.FindAsync(id) ?? throw new NullReferenceException($"Can't find an entity with this Id: {id}.");
```
Comment (MostafaHgz) [inline SchedsForums.Infrastructure/Repositories/Common/GenericRepository.cs]: it's not the responsibility of the repository to throw the error. it should return a nullable value. use `SingleOrDefaultAsync`

File: SchedsForums.Infrastructure/Repositories/MajorRepository.cs [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+using SchedsForums.Infrastructure.Repositories.Common;
+
+namespace SchedsForums.Infrastructure.Repositories
+{
+    public class MajorRepository : GenericRepository<Major>, IMajorRepository
+    {
+        private readonly ForumsDbContext _context;
+        public MajorRepository(ForumsDbContext context) : base(context)
+        {
+            _context = context;
```
Comment (MostafaHgz) [inline SchedsForums.Infrastructure/Repositories/MajorRepository.cs]: remove all repositories that you didn't use in this story as they are not part of the story scope and you didn't use them
always add null checks for the when injecting. also you're injecting the ForumsDbContext inside the Major repository which is incorrect.

File: SchedsForums.Infrastructure/Services/DbContextRegistrationService.cs [hunk 19]
```diff
@@ -0,0 +1,21 @@
+﻿using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.DependencyInjection;
+using SchedsForums.Infrastructure.Contexts;
+
+namespace SchedsForums.Persistence.Services
+{
+    public static class DbContextRegistrationService
+    {
+        public static IServiceCollection AddPersistenceServices(this IServiceCollection services)
```
Comment (MostafaHgz) [inline SchedsForums.Infrastructure/Services/DbContextRegistrationService.cs]: this adds the DB context not a service

File: SchedsForums.Infrastructure/Services/PasswordService.cs [hunk 20]
```diff
@@ -0,0 +1,18 @@
+﻿using SchedsForums.Infrastructure.Services.Interfaces;
+
+namespace SchedsForums.Infrastructure.Services
+{
+    public class PasswordService : IPasswordService
```
Comment (MostafaHgz) [inline SchedsForums.Infrastructure/Services/PasswordService.cs]: what if I want to encrypt or hash any other thing than the password. this should be called `HashingService`

File: SchedsForums.Infrastructure/ServicesRegistration.cs [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace SchedsForums.Infrastructure
+{
+    public static class ServicesRegistration
+    {
+        public static IServiceCollection RegisterInfrastructureRepositories(this IServiceCollection services)
+        {
+            services.AddScoped<IAdminRepository, AdminRepository>();
+            services.AddScoped<IModeratorRepository, ModeratorRepository>();
+            services.AddScoped<IStudentRepository, StudentRepository>();
```
Comment (MostafaHgz) [inline SchedsForums.Infrastructure/ServicesRegistration.cs:25]: only student repository is needed remove the rest

File: SchedsForums.Interface/IBaseService.cs [hunk 22]
```diff
@@ -0,0 +1,6 @@
+﻿namespace SchedsForums.Interface
```
Comment (MostafaHgz) [inline SchedsForums.Interface/IBaseService.cs:1]: this does nothing

File: SchedsForums.Interface/Repositories/IAdminRepository.cs [hunk 23]
```diff
@@ -0,0 +1,8 @@
+﻿using SchedsForums.Domain.Entities.Users;
```
Comment (MostafaHgz) [inline SchedsForums.Interface/Repositories/IAdminRepository.cs:1]: remove all unwanted interfaces

File: SchedsForums.Interface/IGenericRepository.cs [hunk 24]
```diff
@@ -0,0 +1,12 @@
+﻿namespace SchedsForums.Interface
+{
+    public interface IGenericRepository<T> where T : IBaseEntity
+    {
+        Task<T> GetByIdAsync(string id);
+        Task<IEnumerable<T>> GetAllAsync();
+        Task<T> InsertAsync(T entity);
+        Task<T> UpdateAsync(T entity);
+        Task<T> DeleteAsync(string id);
+
```
Comment (MostafaHgz) [inline SchedsForums.Interface/IGenericRepository.cs:10]: remove extra line

File: SchedsForums.Persistence/Contexts/ForumsDbContext.cs [hunk 25]
```diff
@@ -0,0 +1,32 @@
+﻿using Microsoft.EntityFrameworkCore;
+using SchedsForums.Domain.Entities;
+using SchedsForums.Domain.Entities.Common;
+using SchedsForums.Domain.Entities.Users;
+namespace SchedsForums.Persistence.Contexts
+{
+    public class ForumsDbContext : DbContext
```
Comment (MostafaHgz) [inline SchedsForums.Persistence/Contexts/ForumsDbContext.cs:7]: rename this to the name of the application

File: SchedsForums.Infrastructure/Contexts/ForumsDbContext.cs [hunk 26]
```diff
@@ -0,0 +1,32 @@
+﻿using Microsoft.EntityFrameworkCore;
+using SchedsForums.Domain.Entities;
+using SchedsForums.Domain.Entities.Common;
+using SchedsForums.Domain.Entities.Users;
+namespace SchedsForums.Infrastructure.Contexts
+{
+    public class ForumsDbContext : DbContext
+    {
+        public ForumsDbContext(DbContextOptions<ForumsDbContext> options) : base(options)
```
Comment (MostafaHgz) [inline SchedsForums.Infrastructure/Contexts/ForumsDbContext.cs]: you have this file twice

File: SchedsForums.Infrastructure/Services/PasswordService.cs [hunk 27]
```diff
@@ -0,0 +1,18 @@
...
+        public string HashPassword(string password)
+        {
+            return BCrypt.Net.BCrypt.HashPassword(password);
+        }
+
+        public bool VerifyPassword(string password, string hash)
+        {
+            return BCrypt.Net.BCrypt.Verify(password, hash);
+        }
+
```
Comment (MostafaHgz) [inline SchedsForums.Infrastructure/Services/PasswordService.cs]: remove extra line

File: SchedsForums.Persistence/Contexts/ForumsDbContext.cs (hunk 25 above)
Comment (MostafaHgz) [inline SchedsForums.Persistence/Contexts/ForumsDbContext.cs:7]: you have this file twice

File: SchedsForums.Domain/Entities/Major.cs [hunk 28]
```diff
@@ -0,0 +1,8 @@
+﻿using SchedsForums.Domain.Entities.Common;
```
Comment (MostafaHgz) [inline SchedsForums.Domain/Entities/Major.cs]: only the user entity is related to the story's scope. remove all other entities

File: SchedsForums.Domain/Exceptions/EntityNotFoundException.cs [hunk 29]
```diff
@@ -0,0 +1,6 @@
+﻿namespace SchedsForums.Domain.Exceptions
```
Comment (MostafaHgz) [inline SchedsForums.Domain/Exceptions/EntityNotFoundException.cs]: remove this exception and use .NET Exceptions

File: SchedsForums.Domain/Interfaces/IGenericRepository.cs (hunk 7 above)
Comment (MostafaHgz) [reply SchedsForums.Domain/Interfaces/IGenericRepository.cs]: also all interfaces should be in persistance. if you can't do it due to project dependencies we can check it when we next meet

File: SchedsForums.Api/Controllers/StudentController.cs [hunk 30]
```diff
@@ -0,0 +1,28 @@
...
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> CreateStudent([FromBody] CreateStudentDTO studentDto)
+        {
+            var createStudentCommand = new CreateStudentCommand { Student = studentDto };
+
+            var result = await _mediator.Send(createStudentCommand);
+            return Ok(result);
+            //return CreatedAtAction(nameof(GetStudentById), new { id = result.Id }, result);
```
Comment (MostafaHgz) [inline SchedsForums.Api/Controllers/StudentController.cs]: remove commented line

File: SchedsForums.Api/Controllers/StudentController.cs [hunk 31]
```diff
@@ -0,0 +1,28 @@
...
+
+        public StudentsController(IMediator mediator)
+        {
+            _mediator = mediator;
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> CreateStudent([FromBody] CreateStudentDTO studentDto)
+        {
+            var createStudentCommand = new CreateStudentCommand { Student = studentDto };
```
Comment (MostafaHgz) [inline SchedsForums.Api/Controllers/StudentController.cs]: you should make the body of the request is the command itself without constructing a new one

File: SchedsForums.Application/Commands/Students/Create/CreateStudentCommand.cs [hunk 32]
```diff
@@ -0,0 +1,10 @@
+﻿using MediatR;
+using SchedsForums.Application.BaseDTOs;
+
+namespace SchedsForums.Application.Commands.Students.Create
+{
+    public class CreateStudentCommand : IRequest<StudentDTO>
+    {
+        public CreateStudentDTO Student { get; set; }
```
Comment (MostafaHgz) [inline SchedsForums.Application/Commands/Students/Create/CreateStudentCommand.cs]: student dto should be the command. it doesn't make sense to have a field inside a class that has a class that has the data. just make the command have the data you want directly

File: SchedsForums.Application/BaseDTOs/StudentDTO.cs [hunk 33]
```diff
@@ -0,0 +1,8 @@
+﻿namespace SchedsForums.Application.BaseDTOs
+{
+    public class StudentDTO
```
Comment (MostafaHgz) [inline SchedsForums.Application/BaseDTOs/StudentDTO.cs]: is this a request or response. you need to specify

**Review by MostafaHgz — CHANGES_REQUESTED** [review]

File: SchedsForums.Interface/IBaseService.cs (hunk 22 above)
Comment (aabdoo23) [reply SchedsForums.Interface/IBaseService.cs:1]: yet

File: SchedsForums.Infrastructure/Contexts/ForumsDbContext.cs (hunk 26 above)
Comment (aabdoo23) [reply SchedsForums.Infrastructure/Contexts/ForumsDbContext.cs]: i screwed with git, its not wallahy

File: SchedsForums.Infrastructure/Repositories/Common/GenericRepository.cs (hunk 17 above)
Comment (aabdoo23) [reply SchedsForums.Infrastructure/Repositories/Common/GenericRepository.cs]: why singleOrDefault, im using findasync base on this:
image

File: SchedsForums.Infrastructure/Repositories/MajorRepository.cs (hunk 18 above)
Comment (aabdoo23) [reply SchedsForums.Infrastructure/Repositories/MajorRepository.cs]: I did remove the repo as it's not a part of the story ok.
but Im referencing the DbContext in the repo, why wrong though

File: SchedsForums.Domain/Interfaces/IGenericRepository.cs (hunk 7 above)
Comment (aabdoo23) [reply SchedsForums.Domain/Interfaces/IGenericRepository.cs]: ok

**Pull request merged**